src/devices/machine/pla.cpp

```diff
@@ -135,11 +139,21 @@ void pla_device::parse_fusemap()
 
 		for (int i = 0; i < m_inputs; i++)
 		{
-			// complement
-			term->and_mask |= (uint64_t)jed_get_fuse(&jed, fusenum++) << (i + 32);
-
-			// true
-			term->and_mask |= (uint64_t)jed_get_fuse(&jed, fusenum++) << i;
+			if (jed.binfmt == MAXLOADER)
+			{
+				term->and_mask |= (uint64_t)jed_get_fuse(&jed, fusenum++) << (i + 32); // complement
+				term->and_mask |= (uint64_t)jed_get_fuse(&jed, fusenum++) << i;        // true
+			}
+			else if (jed.binfmt == DATAIO)
+			{
+				term->and_mask |= (uint64_t)jed_get_fuse(&jed, fusenum++) << i;        // true
+				term->and_mask |= (uint64_t)jed_get_fuse(&jed, fusenum++) << (i + 32); // complement
+			}
+			else
+			{
+				logerror("PLA parser: Unknown JED format\n");
+				break;
+			}
```

Comment:
If you need additional code in the client, you have a leaky abstraction.

Reply:
What is the client here, I only added an alternative way to unpack the binary format based on a detection in the underlying loader abstraction, did it make it leak more? The alternative is to rewrite the loader level to rearrange the binary to look like a uniformed format. I think that is more likely to break things, this is quite safe imho.

Reply:
The [jedbin_parse()](https://github.com/mamedev/mame/blob/d0399db9b544440a8c687203b560b53375635bbf/src/lib/util/jedparse.cpp#L375) does not actually *parse* the binary data, just the header and copies the binary data to the caller through the jed_data structure. The change just tries t recognise the format, the DataIO format header is number of bytes two times it seems, rather then number of fuses, I haven't found formal documentation of the format though. 

[Jedutil](https://github.com/mamedev/mame/blob/d0399db9b544440a8c687203b560b53375635bbf/src/tools/jedutil.cpp#L7425) just loads a maxloader binary blob from the file or generates it from a .jed file, I added a check to avoid converting DataIO PLA binaries, which probably failed or passed undetected before resulting in unusable results. 

[pla.cpp](https://github.com/mamedev/mame/blob/d0399db9b544440a8c687203b560b53375635bbf/src/devices/machine/pla.cpp#L97) actually parses it so the leaky abstraction is already there, I just added a flag to interpret it differently so that the DataIO format can work.

Reply:
`pla.cpp` is the client in this case.  You've screwed up the `.jed` so it's no longer a format as such, it's a wrapper for different kinds of blobs.  That's completely unacceptable.

Reply:
Got it, I didn't want to touch the utility but the way the PLA loader works I had to. There is no way to tell what format a PLA binary has unless you know so I think many people has tried to load the wrong binary format and failed and not used it. I just convert the binary by hand then and comment that in the code and remove the PLA changes